src/main/scala/firrtl/passes/wiring/Wiring.scala

```diff
@@ -10,157 +10,159 @@ import firrtl.Mappers._
-      val total = countInstances(childrenMap, c.main, m)
-      val nTop = countInstances(childrenMap, c.main, wi.top)
-      val perTop = countInstances(childrenMap, wi.top, m)
-      if(total != nTop * perTop)
-        throw new WiringException(s"Module ${wi.top} does not contain all instances of $m.")
-    }
+    // Split out wiring info
+    val (ComponentName(compName, ModuleName(source,_)), sinks, pin) = (wi.source, wi.sinks, wi.pin)
+    val iGraph = new InstanceGraph(c)
 
     // Create valid port names for wiring that have no name conflicts
     val portNames = c.modules.foldLeft(Map.empty[String, String]) { (map, m) =>
       map + (m.name -> {
         val ns = Namespace(m)
-        if(sinks.contains(m.name)) ns.newName(pin)
+        if(sinks.filter(namedName(_) == m.name).nonEmpty) ns.newName(pin)
         else ns.newName(tokenize(compName) filterNot ("[]." contains _) mkString "_")
       })
     }
 
-    // Create a lineage tree from children map
-    val lineages = getLineage(childrenMap, wi.top)
-
-    // Populate lineage tree with relationship information, i.e. who is source,
-    //   sink, parent of source, etc.
-    val withFields = setSharedParent(wi.top)(setFields(sinks, source)(lineages))
-
-    // Populate lineage tree with what to instantiate, connect to/from, etc.
-    val withThings = setThings(portNames, compName)(withFields)
-
-    // Create a map from module name to lineage information
-    val map = pointToLineage(withThings)
-
     // Obtain the source component type
     val sourceComponentType = getType(c, source, compName)
+    val sinkComponents = sinks
+      .foldLeft(Map.empty[String, String]) { (map, m) =>
+        map ++ { m match {
+          case ComponentName(c, ModuleName(m, _)) => Map(m -> c)
+          case _ => Map()
+        }}}
+
+    // Determine "ownership" of sources to sinks via minimum distance
+    val owners = sinksToSources(sinks, source, iGraph)
+
+    // Determine port and pending modifications (Metadata) for all
+    // sink--source ownership pairs
+    val meta = new mutable.HashMap[String, Metadata].withDefaultValue(Metadata())
+    owners.map { case (sink, source) =>
+      val lca = iGraph.lowestCommonAncestor(sink, source)
+
+      // Compute metadata along Sink to LCA paths
+      sink.drop(lca.size - 1).sliding(2).toList.reverse.map {
+        case Seq(WDefInstance(_,_,pm,_), WDefInstance(_,ci,cm,_)) =>
+          val to = s"$ci.${portNames(cm)}"
+          val from = s"${portNames(pm)}"
+          meta(pm) = meta(pm).copy(
+            addPort = Some((portNames(pm), DecWire)),
+            cons = (meta(pm).cons :+ (to, from)).distinct
+          )
+          meta(cm) = meta(cm).copy(
+            addPort = Some((portNames(cm), DecInput))
+          )
+        case _ =>
+          throw new WiringException(s"""Unexpectedly short path from LCA to sink (possibly due to source being a submodule of a sink?):
+  - LCA: ${lca.map(_.name)}
+  - sink: ${sink.map(_.name)}""")
```

Comment:
I would prefer these tabbed over a bit with some tactical use of `|` and stripMargin